File: bindings/zig/src/link.zig
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

const c = @import("c.zig");
const types = @import("types.zig");
const util = @import("util.zig");

pub fn open(node: types.Node, dest_hash: []const u8) types.Error!types.Link {
    if (dest_hash.len != types.hash_len) return error.InvalidArg;
    const h = c.rns_link_open(types.asU64(node), dest_hash.ptr);
    if (h == 0) return error.Internal;
    return @enumFromInt(h);
}

pub fn send(link: types.Link, data: []const u8) types.Error!void {
    if (data.len == 0) return error.InvalidArg;
    try types.mapCode(c.rns_link_send(types.asU64(link), data.ptr, data.len));
}

pub fn sendResource(link: types.Link, data: []const u8, name: []const u8) types.Error!void {
    var name_buf: [256]u8 = undefined;
    const name_z = if (name.len == 0) null else try util.requireCString(name, &name_buf);
    try types.mapCode(c.rns_link_send_resource(
        types.asU64(link),
        if (data.len == 0) null else data.ptr,
        data.len,
        name_z,
    ));
}

pub fn close(link: types.Link) types.Error!void {
    try types.mapCode(c.rns_link_close(types.asU64(link)));
}

pub fn id(link: types.Link) types.Error![types.hash_len]u8 {
    var out: [types.hash_len]u8 = undefined;
    var written: usize = 0;
    try types.mapCode(c.rns_link_id(types.asU64(link), &out, out.len, &written));
    if (written != types.hash_len) return error.Truncated;
    return out;
}

pub fn request(
    node: types.Node,
    link: types.Link,
    path: []const u8,
    data: []const u8,
    timeout_ms: c_int,
) types.Error![types.hash_len]u8 {
    var path_buf: [256]u8 = undefined;
    const path_z = try util.requireCString(path, &path_buf);
    var request_id: [types.hash_len]u8 = undefined;
    var written: usize = 0;
    try types.mapCode(c.rns_link_request(
        types.asU64(node),
        types.asU64(link),
        path_z,
        if (data.len == 0) null else data.ptr,
        data.len,
        timeout_ms,
        &request_id,
        request_id.len,
        &written,
    ));
    if (written != types.hash_len) return error.Truncated;
    return request_id;
}

pub fn respond(node: types.Node, request_id: []const u8, data: []const u8) types.Error!void {
    if (request_id.len == 0) return error.InvalidArg;
    try types.mapCode(c.rns_request_respond(
        types.asU64(node),
        request_id.ptr,
        request_id.len,
        if (data.len == 0) null else data.ptr,
        data.len,
    ));
}

pub fn respondFile(node: types.Node, request_id: []const u8, filename: []const u8, data: []const u8) types.Error!void {
    if (request_id.len == 0 or filename.len == 0) return error.InvalidArg;
    var name_buf: [256]u8 = undefined;
    const name_z = try util.requireCString(filename, &name_buf);
    try types.mapCode(c.rns_request_respond_file(
        types.asU64(node),
        request_id.ptr,
        request_id.len,
        name_z,
        if (data.len == 0) null else data.ptr,
        data.len,
    ));
}

File: bindings/rust/src/link.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

use std::ffi::CString;
use std::ptr;

use crate::error::{map_code, Error, Result};
use crate::ffi::{self, HASH_LEN};
use crate::node::Node;

pub struct Link {
    handle: u64,
}

impl Link {
    pub fn open(node: &Node, dest_hash: &[u8; HASH_LEN]) -> Result<Self> {
        let h = unsafe { ffi::rns_link_open(node.handle(), dest_hash.as_ptr()) };
        if h == 0 {
            return Err(Error::Internal);
        }
        Ok(Self { handle: h })
    }

    pub fn send(&self, data: &[u8]) -> Result<()> {
        if data.is_empty() {
            return Err(Error::InvalidArg);
        }
        map_code(unsafe { ffi::rns_link_send(self.handle, data.as_ptr(), data.len()) })
    }

    pub fn send_resource(&self, data: &[u8], name: &str) -> Result<()> {
        let name = if name.is_empty() {
            None
        } else {
            Some(CString::new(name).map_err(|_| Error::InvalidArg)?)
        };
        map_code(unsafe {
            ffi::rns_link_send_resource(
                self.handle,
                if data.is_empty() {
                    ptr::null()
                } else {
                    data.as_ptr()
                },
                data.len(),
                name.as_ref().map_or(ptr::null(), |n| n.as_ptr()),
            )
        })
    }

    pub fn close(&self) -> Result<()> {
        map_code(unsafe { ffi::rns_link_close(self.handle) })
    }

    pub fn id(&self) -> Result<[u8; HASH_LEN]> {
        let mut out = [0u8; HASH_LEN];
        let mut written = 0usize;
        map_code(unsafe {
            ffi::rns_link_id(self.handle, out.as_mut_ptr(), out.len(), &mut written)
        })?;
        if written != HASH_LEN {
            return Err(Error::Truncated);
        }
        Ok(out)
    }

    pub fn request(
        &self,
        node: &Node,
        path: &str,
        data: &[u8],
        timeout_ms: i32,
    ) -> Result<[u8; HASH_LEN]> {
        if path.is_empty() {
            return Err(Error::InvalidArg);
        }
        let path = CString::new(path).map_err(|_| Error::InvalidArg)?;
        let mut request_id = [0u8; HASH_LEN];
        let mut written = 0usize;
        map_code(unsafe {
            ffi::rns_link_request(
                node.handle(),
                self.handle,
                path.as_ptr(),
                if data.is_empty() {
                    ptr::null()
                } else {
                    data.as_ptr()
                },
                data.len(),
                timeout_ms,
                request_id.as_mut_ptr(),
                request_id.len(),
                &mut written,
            )
        })?;
        if written != HASH_LEN {
            return Err(Error::Truncated);
        }
        Ok(request_id)
    }

    pub fn handle(&self) -> u64 {
        self.handle
    }
}

impl Drop for Link {
    fn drop(&mut self) {
        if self.handle != 0 {
            unsafe {
                let _ = ffi::rns_link_close(self.handle);
            }
            self.handle = 0;
        }
    }
}

pub fn request_respond(node: &Node, request_id: &[u8], data: &[u8]) -> Result<()> {
    if request_id.is_empty() {
        return Err(Error::InvalidArg);
    }
    map_code(unsafe {
        ffi::rns_request_respond(
            node.handle(),
            request_id.as_ptr(),
            request_id.len(),
            if data.is_empty() {
                ptr::null()
            } else {
                data.as_ptr()
            },
            data.len(),
        )
    })
}

pub fn request_respond_file(
    node: &Node,
    request_id: &[u8],
    filename: &str,
    data: &[u8],
) -> Result<()> {
    if request_id.is_empty() || filename.is_empty() {
        return Err(Error::InvalidArg);
    }
    let filename = CString::new(filename).map_err(|_| Error::InvalidArg)?;
    map_code(unsafe {
        ffi::rns_request_respond_file(
            node.handle(),
            request_id.as_ptr(),
            request_id.len(),
            filename.as_ptr(),
            if data.is_empty() {
                ptr::null()
            } else {
                data.as_ptr()
            },
            data.len(),
        )
    })
}
